undistort_example.py

```diff
@@ -0,0 +1,139 @@
+import open3d as o3d
+import argparse
+import numpy as np
+import mrob
+import math
+
+from pathlib import Path
+from scipy.spatial.transform import Rotation as R
+
+
+def get_intermediate_transform(initial_pose, target_pose, coef):
+    dxi = mrob.geometry.SE3(target_pose @ np.linalg.inv(initial_pose)).Ln()
+    return mrob.geometry.SE3(coef * dxi).T() @ initial_pose
+
+def cut_segment(pcd, theta_min, theta_max, coef):
+    # Convert the point cloud points to a numpy array
+    points = np.asarray(pcd.points)
+    thetas = np.asarray([math.atan2(p[1],p[0]) for p in points])
+
+    # Create a mask for points that fall within the ROI
+    mask_roi = (thetas < theta_max) & (thetas >= theta_min)
+    trues = mask_roi.sum()
+
+    new_pcd = o3d.geometry.PointCloud()
+    if trues > 0:
+        new_colors = np.full((trues, 3), [0., 0., coef])
+        new_points = points[mask_roi]
+
+        new_pcd.points = o3d.utility.Vector3dVector(new_points)
+        new_pcd.colors = o3d.utility.Vector3dVector(new_colors)
+
+    return new_pcd
+
+
+def undistort_cloud(pcd_initial, pcd_target, initial_pose, target_pose, N_segments):
+    N_segments += 1 # starting separations from 0
+    undistort_result = []
+    coefs = np.linspace(0, 1, N_segments)
+    start_angle = np.pi 
+    sectors = np.linspace(start_angle, start_angle - 2 * np.pi, N_segments)
+
+    # applying poses to pcd
+    pcd_initial.transform(pose_start)
+    pcd_target.transform(pose_fin)
+    
+    # calculating transform between two poses
+    initial_transform = np.eye(4, 4)
+    reg_p2p = o3d.pipelines.registration.registration_icp(
+        pcd_initial, pcd_target, 0.1, initial_transform,
+        o3d.pipelines.registration.TransformationEstimationPointToPoint(),  
+        o3d.pipelines.registration.ICPConvergenceCriteria(max_iteration=20000),
+    )
+
+    # returning initial pcd to it's original state
+    # this is done for correct sector cutting
+    pcd_initial.transform(np.linalg.inv(pose_start))
+
+
+    coefs = np.roll(coefs, -len(coefs)//2 ) # rolled due to lidar ray start
+    for i in range(1, N_segments):
+        # cutting a segment
+        segment = cut_segment(pcd_initial, sectors[i], start_angle,  coefs[i-1])
+
+        # getting composite transform matrix: initial -> pose --(gradual)--> target_pcd pose
+        transform_matrix = get_intermediate_transform(initial_pose, reg_p2p.transformation@pose_start, coefs[i-1])
+
+        # applying transform to the segment
+        undistort_result.append(segment.transform(transform_matrix))
+        
+        start_angle = sectors[i]
+
+    return undistort_result
+
+def read_pose(filename):
+    matrix = np.eye(4)
+    try:
+        f = open(filename, newline='')
+    except OSError:
+        print(f"Could not open/read file: {filename}. Default pose returned", filename)
+    with f:
```

Comment:
https://www.freecodecamp.org/news/with-open-in-python-with-statement-syntax-example/